datafusion/substrait/src/logical_plan/producer.rs

```diff
@@ -999,449 +1305,421 @@ pub fn make_binary_op_scalar_func(
-            arg,
-            schema,
-            col_ref_offset,
-            extensions,
-        ),
-        Expr::IsFalse(arg) => to_substrait_unary_scalar_fn(
-            state,
-            "is_false",
-            arg,
-            schema,
-            col_ref_offset,
-            extensions,
-        ),
-        Expr::IsUnknown(arg) => to_substrait_unary_scalar_fn(
-            state,
-            "is_unknown",
-            arg,
-            schema,
-            col_ref_offset,
-            extensions,
-        ),
-        Expr::IsNotTrue(arg) => to_substrait_unary_scalar_fn(
-            state,
-            "is_not_true",
-            arg,
-            schema,
-            col_ref_offset,
-            extensions,
-        ),
-        Expr::IsNotFalse(arg) => to_substrait_unary_scalar_fn(
-            state,
-            "is_not_false",
-            arg,
-            schema,
-            col_ref_offset,
-            extensions,
-        ),
-        Expr::IsNotUnknown(arg) => to_substrait_unary_scalar_fn(
-            state,
-            "is_not_unknown",
-            arg,
-            schema,
-            col_ref_offset,
-            extensions,
-        ),
-        Expr::Negative(arg) => to_substrait_unary_scalar_fn(
-            state,
-            "negate",
-            arg,
-            schema,
-            col_ref_offset,
-            extensions,
-        ),
-        _ => {
-            not_impl_err!("Unsupported expression: {expr:?}")
-        }
+                ),
+            },
+        ))),
+    };
+    if *negated {
+        let function_anchor = producer.register_function("not".to_string());
+
+        #[allow(deprecated)]
+        Ok(Expression {
+            rex_type: Some(RexType::ScalarFunction(ScalarFunction {
+                function_reference: function_anchor,
+                arguments: vec![FunctionArgument {
+                    arg_type: Some(ArgType::Value(substrait_subquery)),
+                }],
+                output_type: None,
+                args: vec![],
+                options: vec![],
+            })),
+        })
+    } else {
+        Ok(substrait_subquery)
     }
 }
 
+pub fn from_unary_expr(
+    producer: &mut impl SubstraitProducer,
+    expr: &Expr,
+    schema: &DFSchemaRef,
+) -> Result<Expression> {
+    let (fn_name, arg) = match expr {
+        Expr::Not(arg) => ("not", arg),
+        Expr::IsNull(arg) => ("is_null", arg),
+        Expr::IsNotNull(arg) => ("is_not_null", arg),
+        Expr::IsTrue(arg) => ("is_true", arg),
+        Expr::IsFalse(arg) => ("is_false", arg),
+        Expr::IsUnknown(arg) => ("is_unknown", arg),
+        Expr::IsNotTrue(arg) => ("is_not_true", arg),
+        Expr::IsNotFalse(arg) => ("is_not_false", arg),
+        Expr::IsNotUnknown(arg) => ("is_not_unknown", arg),
+        Expr::Negative(arg) => ("negate", arg),
+        expr => not_impl_err!("Unsupported expression: {expr:?}")?,
+    };
+    to_substrait_unary_scalar_fn(producer, fn_name, arg, schema)
```

Comment:
Consolidated the handling of unary expression like Not, IsNull, IsNotNull etc into a single function for improved readability.